src/data/arena-tag.ts

```diff
@@ -913,6 +913,121 @@ export class DelayedAttackTag extends ArenaTag {
+   */
+  public addHeal(targetIndex: BattlerIndex, healEffect: PendingHealEffect): void {
+    if (this.pendingHeals.has(targetIndex)) {
+      const existingHealEffects = this.pendingHeals.get(targetIndex);
+      if (existingHealEffects?.every((he) => he.moveId !== healEffect.moveId)) {
+        existingHealEffects.push(healEffect);
+      }
+    } else {
+      this.pendingHeals.set(targetIndex, [healEffect]);
+    }
+  }
+
+  /** Removes default on-remove message */
+  override onRemove(_arena: Arena): void {}
+
+  /** This arena tag is removed at the end of the turn if no pending healing effects are on the field */
+  override lapse(_arena: Arena): boolean {
+    return this.pendingHeals.size > 0;
+  }
+
+  /**
+   * Applies a pending healing effect on the given target index. If an effect is found for
+   * the index, the Pokemon at that index is healed to full HP, is cured of any non-volatile status,
+   * and has its PP fully restored (if the effect is from Lunar Dance).
+   * @param arena - The {@linkcode Arena} containing this tag
+   * @param simulated - If `true`, suppresses changes to game state
+   * @param pokemon - The {@linkcode Pokemon} receiving the healing effect
+   * @returns `true` if the target Pokemon was healed by this effect
+   * @todo This should also be called when a Pokemon moves into a new position via Ally Switch
+   */
+  override apply(arena: Arena, simulated: boolean, pokemon: Pokemon): boolean {
+    const targetIndex = pokemon.getBattlerIndex();
+    const targetEffects = this.pendingHeals.get(targetIndex);
+
+    if (simulated) {
+      return !!targetEffects?.length;
+    }
+
+    const healEffect = targetEffects?.find((effect) => this.canApply(effect, pokemon));
+    if (targetEffects && healEffect) {
+      const { sourceId, moveId, restorePP, healMessageKey } = healEffect;
+      const sourcePokemon = globalScene.getPokemonById(sourceId);
+      if (!sourcePokemon) {
+        console.warn(`Source of pending ${allMoves.get(moveId).name} effect is undefined!`);
+        targetEffects.splice(targetEffects.indexOf(healEffect), 1);
+        // Re-evaluate after the invalid heal effect is removed
+        return this.apply(arena, simulated, pokemon);
+      }
+
+      globalScene.phaseManager.queuePokemonHealPhase(true, targetIndex, pokemon.getMaxHp(), {
+        message: i18next.t(healMessageKey, { pokemonName: getPokemonNameWithAffix(sourcePokemon) }),
+        healStatus: true,
+        fullRestorePP: restorePP,
+      });
+
+      targetEffects.splice(targetEffects.indexOf(healEffect), 1);
+    }
+
+    if (targetEffects?.length === 0) {
+      this.pendingHeals.delete(targetIndex);
+    }
+
+    return !isNullOrUndefined(healEffect);
+  }
+
+  /**
+   * Determines if the given {@linkcode PendingHealEffect} can immediately heal
+   * the given target {@linkcode Pokemon}.
+   * @param healEffect the {@linkcode PendingHealEffect} to evaluate
+   * @param pokemon the {@linkcode Pokemon} to evaluate against
```

Comment:
NIT: I recently learned the dashes have to put manually. It's a mistake from TS that they get addded automatically in the peek.


```suggestion
   * @param healEffect - The {@linkcode PendingHealEffect} to evaluate
   * @param pokemon - The {@linkcode Pokemon} to evaluate against
```